src/main/java/org/candlepin/service/impl/PromotedContent.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2009 - 2021 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.service.impl;
+
+import org.candlepin.model.Environment;
+import org.candlepin.model.EnvironmentContent;
+import org.candlepin.model.ProductContent;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * A collection of content promoted to environments.
+ */
+public class PromotedContent {
```

Comment:
This class is a glorified wrapper around a hash map, which is fine, but we then use a factory method to utilize it rather than anything that would better fit its property of a wrapper, such as a fluent, builder-style API. This would keep the construction in context, which is implied to be intended by the first log message in the builder function.

```
PromotedContent promotedContent = new PromotedContent()
    .from(environment)
    .from(environment);
```

While I understand this design likely came about because of how we use it today (that is, we get the list of environments from the curator and throw it at the `.from` method), I don't think we should retain design choices from the past if we're going to spend the effort making a new abstraction layer to replace them.

This also eliminates the need for multiple constructors, and fully encapsulates the underlying map.

As a side note, the example above uses a single environment mutator since the order in which environments are added is critical to how this operates. That should be front-and-center with the design here, since it's so central to how this object is going to be used. A `.fromAll` could also work, but it'd need to be limited to order-retaining collections for that very reason.

Reply:
Done.